File: functions/src/domain/providers/functions/FunctionsUseCaseProvider.ts
/**
 * Functions use case provider
 * Groups Firebase Functions related use cases
 */

import { FunctionsRepository, RepositoryFactory } from '../../repositories';
import { FunctionsRepositoryFactoryContext } from '../../repositories';
import { HelloWorldUseCase } from '../../usecases/functions/HelloWorldUseCase';

export interface FunctionsUseCases {
  readonly helloWorldUseCase: HelloWorldUseCase;
  readonly functionsRepository: FunctionsRepository;
}

export class FunctionsUseCaseProvider {
  constructor(
    private readonly functionsRepositoryFactory: RepositoryFactory<FunctionsRepositoryFactoryContext, FunctionsRepository>
  ) {}

  create(options: {
    functionsContext?: FunctionsRepositoryFactoryContext;
  } = {}): FunctionsUseCases {
    const functionsRepository = this.functionsRepositoryFactory.create(
      options.functionsContext || {}
    );

    return {
      helloWorldUseCase: new HelloWorldUseCase(functionsRepository),
      functionsRepository,
    };
  }
}

File: domain/src/main/java/com/example/domain/provider/functions/FunctionsUseCaseProvider.kt
package com.example.domain.provider.functions

import com.example.domain.repository.base.UserRepository
import com.example.domain.usecase.functions.HelloWorldUseCase
import com.example.domain.usecase.functions.HelloWorldUseCaseImpl
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.jvm.JvmSuppressWildcards

/**
 * Firebase Functions 관련 UseCase들을 제공하는 Provider
 * Clean Architecture의 UseCase Provider 패턴을 따릅니다.
 */
data class FunctionsUseCases(
    val helloWorldUseCase: HelloWorldUseCase,
    val userRepository: UserRepository
)

@Singleton
class FunctionsUseCaseProvider @Inject constructor(
    private val userRepository: UserRepository
) {
    
    /**
     * Firebase Functions 관련 UseCase들을 생성하고 반환합니다.
     * 
     * @return FunctionsUseCases 객체
     */
    fun create(): FunctionsUseCases {
        return FunctionsUseCases(
            helloWorldUseCase = HelloWorldUseCaseImpl(userRepository),
            userRepository = userRepository
        )
    }
}